compiler/src/iree/compiler/Codegen/Common/LinearizeMemRefs.cpp

```diff
@@ -0,0 +1,343 @@
+                                   MemRefType &newMemrefType) {
+  // Fetch linearized shape.
+  // TODO(avarma): Take into account different src/dst bits.
+  int srcBits = memrefType.getElementType().getIntOrFloatBitWidth();
+  SmallVector<int64_t> linearizedShape =
+      getLinearizedShape(memrefType, srcBits, srcBits);
+  // Fetch offset and strides of the old memref.
+  SmallVector<int64_t> strides;
+  int64_t offset;
+  if (failed(getStridesAndOffset(memrefType, strides, offset)))
+    return failure();
+  if (!strides.empty() && strides.back() != 1)
+    return failure();
+  // Form layout for the linearized memref.
+  StridedLayoutAttr layoutAttr;
+  // If the offset is 0, we do not need a strided layout as the stride is
+  // 1, so we only use the strided layout if the offset is not 0.
+  if (offset != 0) {
+    layoutAttr = StridedLayoutAttr::get(memrefType.getContext(), offset,
+                                        ArrayRef<int64_t>{1});
+  }
+  Type elementType = memrefType.getElementType();
+  newMemrefType = MemRefType::get(linearizedShape, elementType, layoutAttr,
+                                  memrefType.getMemorySpace());
+  return success();
+}
+
+static LogicalResult
+getLinearizedTypeFromSourceType(MemRefType currentTypeOfSourceMemref,
+                                MemRefType &linearizedType) {
+  if (!currentTypeOfSourceMemref)
+    return failure();
+  if (currentTypeOfSourceMemref.getRank() < 2)
+    return success();
+  // Convert current type later.
+  return linearizeType(currentTypeOfSourceMemref, linearizedType);
+}
+
+template <typename OpTy>
+struct LinearizeMemrefAlloc : public OpRewritePattern<OpTy> {
+  LinearizeMemrefAlloc(MLIRContext *context, PatternBenefit benefit = 10)
+      : OpRewritePattern<OpTy>(context, benefit) {}
+
+  LogicalResult matchAndRewrite(OpTy allocOp,
+                                PatternRewriter &rewriter) const override {
+    static_assert(std::is_same<OpTy, memref::AllocOp>() ||
+                      std::is_same<OpTy, memref::AllocaOp>(),
+                  "expected only memref::AllocOp or memref::AllocaOp");
+    Location loc = allocOp->getLoc();
+    MemRefType currentTypeOfSourceMemref =
+        dyn_cast<MemRefType>(allocOp.getMemref().getType());
+    MemRefType newTypeOfSourceMemref;
+    if (failed(getLinearizedTypeFromSourceType(currentTypeOfSourceMemref,
+                                               newTypeOfSourceMemref))) {
+      return failure();
+    }
+    if (currentTypeOfSourceMemref.getRank() < 2)
+      return success();
+
+    auto elementType = currentTypeOfSourceMemref.getElementType();
+    int srcBits = elementType.getIntOrFloatBitWidth();
+
+    OpFoldResult zero = rewriter.getIndexAttr(0);
+
+    // Get linearized type.
+    int dstBits = srcBits;
+    SmallVector<OpFoldResult> sizes = allocOp.getMixedSizes();
+
+    memref::LinearizedMemRefInfo linearizedMemRefInfo =
+        memref::getLinearizedMemRefOffsetAndSize(
```

Comment:
I thought we prefered using `affine.linearize_index`. IIRC this method uses `affine.apply` instead.